xlators/cluster/dht/src/dht-rebalance.c

```diff
@@ -994,6 +994,49 @@ __dht_check_free_space(xlator_t *this, xlator_t *to, xlator_t *from, loc_t *loc,
     return ret;
 }
 
+static int32_t
+dht_rebalance_sparse_segment(xlator_t *subvol, fd_t *fd, off_t *offset,
+                             size_t *size)
+{
+    off_t data;
+    off_t hole;
+    int32_t ret;
+
+    do {
+        data = *offset;
+        ret = syncop_seek(subvol, fd, data, GF_SEEK_DATA, NULL, &data);
+        if (ret < 0) {
+            if (ret == -ENXIO) {
+                ret = 0; /* No more data segments */
+            }
+            return ret; /* Error occurred */
+        }
+
+        /* Starting at the offset of the last data segment, find the
+         * next hole */
+        ret = syncop_seek(subvol, fd, data, GF_SEEK_HOLE, NULL, &hole);
+        if (ret < 0) {
+            /* If an error occurred here it's a real error because
+             * if the seek for a data segment was successful then
+             * necessarily another hole must exist (EOF is a hole)
+             */
+            return ret;
+        }
+
+        /* Calculate the total size of the current data block */
+        *size = hole - data;
+
+        /* It could happen that at the same offset we detected data in the
+         * first seek, there could be a hole in the second seek if user is
+         * modifying the file concurrently. In this case we need to find a
+         * new data segment to migrate. */
+    } while (*size <= 0);
```

Comment:
size is size_t, are you sure about this condition?

Reply:
Yes, I already thought about that. A negative value should be impossible in theory, so it seemed ok at first sight. However on second thought, I think it can be made more robust with little changes... I'll do that.